src/pj_obs_api.c

```diff
@@ -222,41 +266,150 @@ PJ_COORD proj_trans_coord (PJ *P, enum proj_direction direction, PJ_COORD coo) {
+        PJ *P = {...};
+        ...
+        proj_transform (
+            P, PJ_INV, sizeof(XYQS),
+            &(survey[0].x), 345,     (*  We have 345 eastings  *)
+            &(survey[0].y), 345,     (*  ...and 345 northings. *)
+            &height, 1,              (*  The height is the constant  23.45 m *)
+            0, 0                     (*  and the time is the constant 0.00 s *)
+        );
+
+    This is similar to the inner workings of the pj_transform function, but the
+    idea of the stride has been generalized to work for any size of basic unit,
+    not just a number of doubles.
+
+**************************************************************************************/
+    PJ_COORD coord;
+    size_t i, nmin;
+    double null_broadcast = 0;
     if (0==P)
-        return HUGE_VAL;
-
-    if (n < 1) {
-        proj_errno_set (P, EINVAL);
-        return HUGE_VAL;
+        return 0;
+
+    /* ignore lengths of null arrays */
+    if (0==x) nx = 0;
+    if (0==y) ny = 0;
+    if (0==z) nz = 0;
+    if (0==t) nt = 0;
+
+    /* and make the nullities point to some real world memory for broadcasting nulls */
+    if (0==nx) x = &null_broadcast;
+    if (0==ny) y = &null_broadcast;
+    if (0==nz) z = &null_broadcast;
+    if (0==nt) t = &null_broadcast;
+    
+    /* nothing to do? */
+    if (0==nx+ny+nz+nt)
+        return 0;
+
+    /* arrays of length 1 are constants, which we broadcast along the longer arrays */
+    /* so we need to find the length of the shortest non-unity array to figure out  */
+    /* how many coordinate pairs we must transform */
+    nmin = (nx > 1)? nx: (ny > 1)? ny: (nz > 1)? nz: (nt > 1)? nt: 1;
+    if ((nx > 1) && (nx < nmin))  nmin = nx;
+    if ((ny > 1) && (ny < nmin))  nmin = ny;
+    if ((nz > 1) && (nz < nmin))  nmin = nz;
+    if ((nt > 1) && (nt < nmin))  nmin = nt;
+
+    /* Check validity of direction flag */
+    switch (direction) {
+        case PJ_FWD:
+        case PJ_INV:
+            break;
+        case PJ_IDENT:
+            return nmin;
+        default:
+            proj_errno_set (P, EINVAL);
+            return 0;
     }
 
-    o.coo = obs.coo;
 
-    for (i = 0;  i < n;  i++) {
-        switch (direction) {
-            case PJ_FWD:
-                u  =  pj_fwdobs (o, P);
-                o  =  pj_invobs (u, P);
-                break;
-            case PJ_INV:
-                u  =  pj_invobs (o, P);
-                o  =  pj_fwdobs (u, P);
-                break;
-            default:
-                proj_errno_set (P, EINVAL);
-                return HUGE_VAL;
+    /* Arrays of length==0 are broadcast as the constant 0     */
+    /* Arrays of length==1 are broadcast as their single value */
+    /* Arrays of length >1 are iterated over (for the first nmin values) */
+    /* The slightly convolved incremental indexing is used due */
+    /* to the stride, which may be any size supported by the platform */
+    for (i = 0; i < nmin; i++) {
+        coord.xyzt.x = *x;
```

Comment:
The alignment will be OK because sizeof(whatever) includes the necessary alignment slack on platforms that need it - otherwise  `foo *p = calloc (300, sizeof(foo);` would not work: `p[0].bar` is situated `sizeof(foo)` bytes from `p[1].bar`, the same is the case for `p[0].baz` and `p[1].baz`.

What we cannot assume is that `sizeof(foo)==sum_of_the_size_of_its_elements`, since there may be slack both between the struct elements, and at the end of the struct (but not at the beginning: The first element of a struct is required to live at the same address as the struct itself).

This is the reason for my comment about "any size supported by the platform", i.e. any size you can get from the sizeof-operator.